{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RNA-seq Pipeline for Detecting Exitrons (Exonic Introns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0. Installs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/commons/home/ncui/miniconda3/envs/vscode/lib/python3.13/site-packages/sorted_nearest/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "from tqdm import tqdm # progress tracker\n",
    "import pyranges as pr # parsing gff\n",
    "import numpy as np\n",
    "import pysam\n",
    "import itertools\n",
    "import pyranges as pr # parsing gff\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Parse Regtools Data \n",
    "- Parse raw junction data from regtools output files\n",
    "\n",
    "- Processes 1 file at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parseJunctionFile(file_path):\n",
    "    # column names for RegTools junction files\n",
    "    regtools_column_names = [\n",
    "        'chrom', 'start_anchor', 'end_anchor', 'name', 'score', 'strand',\n",
    "        'thick_start_orig', 'thick_end_orig', 'item_rgb_orig',\n",
    "        'block_count_orig', 'block_sizes_orig', 'block_starts_orig'\n",
    "    ]\n",
    "    \n",
    "    # extract sample ID from the filename\n",
    "    sample_id = os.path.basename(file_path).split('.')[0]\n",
    "    \n",
    "    # read the file into a pandas DataFrame\n",
    "    df = pd.read_csv(\n",
    "        file_path, sep='\\t', header=None, names=regtools_column_names,\n",
    "        dtype={'chrom': str, 'block_sizes_orig': str, 'block_starts_orig': str}\n",
    "    )\n",
    "        \n",
    "    df['sample_id_source'] = sample_id\n",
    "\n",
    "    # convert relevant columns to numeric types, coercing errors\n",
    "    for col in ['start_anchor', 'end_anchor', 'score']:\n",
    "        if col in df.columns:\n",
    "            df[col] = pd.to_numeric(df[col], errors='coerce')\n",
    "    \n",
    "    # drop rows if info is missing\n",
    "    df.dropna(subset=['start_anchor', 'end_anchor', 'score', 'block_sizes_orig'], inplace=True)\n",
    "    \n",
    "    # ensure int types\n",
    "    for col in ['start_anchor', 'end_anchor', 'score']:\n",
    "        df[col] = df[col].astype(int)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Transform Junction Data\n",
    "- Recalculates junction coordinates, following Regtools documentation to take into account blockSize\n",
    "\n",
    "- Recalculates block size to represent length of junction\n",
    "\n",
    "- Outputs junction info in BED12 format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def transformJunctionData(raw_df):\n",
    "    \n",
    "    # CHROMOSOME FILTERING\n",
    "    original_row_count = len(raw_df)\n",
    "    \n",
    "    # allowed chromosomes\n",
    "    allowed_chrom_numbers = [str(i) for i in range(1, 23)]\n",
    "    allowed_sex_chroms_upper = ['X', 'Y'] \n",
    "    allowed_chromosomes = set()\n",
    "    for num_chrom in allowed_chrom_numbers:\n",
    "        allowed_chromosomes.add(num_chrom)\n",
    "        allowed_chromosomes.add(f\"chr{num_chrom}\")\n",
    "    for sex_chrom in allowed_sex_chroms_upper:\n",
    "        allowed_chromosomes.add(sex_chrom)\n",
    "        allowed_chromosomes.add(sex_chrom.lower())\n",
    "        allowed_chromosomes.add(f\"chr{sex_chrom}\")\n",
    "        allowed_chromosomes.add(f\"chr{sex_chrom.lower()}\")\n",
    "    \n",
    "    raw_df_filtered = raw_df[raw_df['chrom'].isin(allowed_chromosomes)].copy()\n",
    "    filtered_row_count = len(raw_df_filtered)\n",
    "    print(f\"Removed {original_row_count - filtered_row_count} rows with non-standard chromosomes.\")\n",
    "\n",
    "\n",
    "    # JUNCTION COORD CORRECTION\n",
    "    # filter rows for valid blocks\n",
    "    parsed_blocks_list = raw_df_filtered['block_sizes_orig'].str.strip(',').str.split(',')\n",
    "    has_sufficient_blocks = parsed_blocks_list.str.len() >= 2\n",
    "    raw_df_filtered = raw_df_filtered[has_sufficient_blocks].copy()\n",
    "    parsed_blocks_list = parsed_blocks_list[has_sufficient_blocks]\n",
    "    \n",
    "    # recalculating junction coordinates\n",
    "    raw_df_filtered.loc[:, 'overhang_left'] = parsed_blocks_list.str[0].astype(int)\n",
    "    raw_df_filtered.loc[:, 'overhang_right'] = parsed_blocks_list.str[1].astype(int)\n",
    "\n",
    "    junc_start = raw_df_filtered['start_anchor'] + raw_df_filtered['overhang_left']\n",
    "    junc_end = raw_df_filtered['end_anchor'] - raw_df_filtered['overhang_right']\n",
    "\n",
    "    # filter out invalid junctions\n",
    "    valid_junction = junc_start < junc_end\n",
    "    raw_df_filtered = raw_df_filtered[valid_junction].copy()\n",
    "    junc_start = junc_start[valid_junction]\n",
    "    junc_end = junc_end[valid_junction]\n",
    "\n",
    "\n",
    "    junc_length = junc_end - junc_start\n",
    "\n",
    "    # create df\n",
    "    transformed_df = pd.DataFrame()\n",
    "    transformed_df['chrom'] = raw_df_filtered['chrom']\n",
    "    transformed_df['chromStart'] = junc_start\n",
    "    transformed_df['chromEnd'] = junc_end\n",
    "    transformed_df['name'] = raw_df_filtered['name']\n",
    "    transformed_df['score'] = raw_df_filtered['score']\n",
    "    transformed_df['strand'] = raw_df_filtered['strand']\n",
    "    transformed_df['thickStart'] = junc_start\n",
    "    transformed_df['thickEnd'] = junc_end\n",
    "    transformed_df['itemRgb'] = raw_df_filtered['item_rgb_orig']\n",
    "    transformed_df['blockCount'] = 1\n",
    "    transformed_df['blockSizes'] = junc_length.astype(str)\n",
    "    transformed_df['blockStarts'] = \"0\"\n",
    "    transformed_df['sample_id_source'] = raw_df_filtered['sample_id_source']\n",
    "\n",
    "    print(f\"Transformed {len(transformed_df)} junction records.\")\n",
    "    \n",
    "    return transformed_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Find Exitrons Within Junction Data\n",
    "- Converts transformed junction data (transformed_df) and exon data (from gff3 file) into PyRanges objects with labels Chromosome, Start, End, Strand, and Title (a unique junction id formed by chrom:start:end:strand)\n",
    "\n",
    "- Finds junctions that overlap with CDS regions using PyRanges method .overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 903356 CDS regions.\n"
     ]
    }
   ],
   "source": [
    "# convert CDS data to PyRanges object\n",
    "gff = pr.read_gff3(\"gencode.v48.annotation.gff3.gz\")\n",
    "cds = gff[gff.Feature == \"CDS\"]\n",
    "print(f\"Found {len(cds)} CDS regions.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findExitrons(transformed_df):\n",
    "    transformed_df = transformed_df[transformed_df['strand'].isin(['+', '-'])]\n",
    "\n",
    "    # generate a unique ID for each junction (chrom:start:end:strand\n",
    "    unique_id = transformed_df['chrom'].astype(str) + ':' + \\\n",
    "                transformed_df['chromStart'].astype(str) + ':' + \\\n",
    "                transformed_df['chromEnd'].astype(str) + ':' + \\\n",
    "                transformed_df['strand'].astype(str)\n",
    "\n",
    "    # convert junction data to PyRanges object\n",
    "    junction_pr = pr.PyRanges({'Chromosome': transformed_df['chrom'],\n",
    "                    'Start': transformed_df['chromStart'],\n",
    "                    'End': transformed_df['chromEnd'],\n",
    "                    'Strand': transformed_df['strand'],\n",
    "                    'title': unique_id,\n",
    "                    'reads': transformed_df['score'],\n",
    "                    'sourceID': transformed_df['sample_id_source']}) \n",
    "\n",
    "    # find overlapping junctions\n",
    "    contained_junctions = junction_pr.overlap(cds, contained_intervals_only=True, strand_behavior='same')\n",
    "    print(f\"Found {len(contained_junctions)} junctions contained within CDS regions.\")\n",
    "            \n",
    "    return contained_junctions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Compile All Exitron Info\n",
    "- Iterates through each person's file, finding all exitron data then concatenating to a final matrix\n",
    "\n",
    "- Includes sourceID (file name) and junction scores (total reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compileExitronData(directory_path, output_filepath, file_pattern=\"*.bam.junc\"):\n",
    "\n",
    "    all_exitron_info = []\n",
    "    file_paths = glob.glob(os.path.join(directory_path, file_pattern))\n",
    "    print(f\"Found {len(file_paths)} files to process.\")\n",
    "\n",
    "    # testing first 5 out of 100\n",
    "    '''\n",
    "    files_to_process = file_paths[:5]\n",
    "    print(f\"Processing the first {len(files_to_process)} files.\")\n",
    "    '''\n",
    "\n",
    "    for file_path in tqdm(file_paths):\n",
    "        print(\"Parsing new file...\")\n",
    "        file_name_only = os.path.basename(file_path)\n",
    "        try:\n",
    "            # 1.\n",
    "            parsed_data = parseJunctionFile(file_path)\n",
    "            # 2.\n",
    "            transformed_df = transformJunctionData(parsed_data)\n",
    "            # 3.\n",
    "            gr_file = findExitrons(transformed_df)\n",
    "            #4.\n",
    "            all_exitron_info.append(gr_file)\n",
    "\n",
    "        # skip to the next file if an error occurs\n",
    "        except Exception as e:\n",
    "            print(f\"An error occurred while processing file {file_name_only}: {e}\")\n",
    "            import traceback\n",
    "            traceback.print_exc()\n",
    "            continue \n",
    "\n",
    "    # concatenate all individual data into matrix \n",
    "    final_gr = pr.concat(all_exitron_info)\n",
    "    print(f\"\\nSuccessfully compiled exitron data from {len(all_exitron_info)} files.\")\n",
    "    final_gr.to_parquet(output_filepath, index=False)\n",
    "    print(f\"Successfully saved data to {output_filepath}\")\n",
    "    return final_gr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "compileExitronData(\"/gpfs/commons/groups/knowles_lab/atokolyi/als/juncs_min6bp/\", \"/gpfs/commons/home/ncui/project/exitron_data.parquet\", file_pattern=\"*.bam.junc\")\n",
    "# exitron data stored in exitron_data.parquet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Exitron Filtering\n",
    "- Returns wide dataframe of all unique exitrons with sourceIDs as columns and reads as values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filterExitronData(exitron_data_filepath, min_count=30, min_peeps=10):\n",
    "    long_df= pd.read_parquet(exitron_data_filepath)\n",
    "    # drop duplicates created by pyranges .overlap method\n",
    "    long_df.drop_duplicates(subset=['title', 'sourceID'], inplace=True)\n",
    "\n",
    "    # create filtered wide df\n",
    "    wide_df = long_df.pivot(index='title', columns='sourceID', values='reads').fillna(0)\n",
    "\n",
    "    wide_is_gt_count = wide_df > min_count\n",
    "    wide_is_gt_count_sums = wide_is_gt_count.sum(axis=1)\n",
    "    keep_juncs = wide_is_gt_count_sums[wide_is_gt_count_sums >= min_peeps].index\n",
    "    print(f\"Found {len(keep_juncs)} junctions that passed the filter.\")\n",
    "\n",
    "    # construct final matrix\n",
    "    filtered_exitron_data = wide_df.loc[keep_juncs].copy()\n",
    "    filtered_exitron_data = filtered_exitron_data.where(filtered_exitron_data >= min_count, 0)\n",
    "\n",
    "    return filtered_exitron_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/ipykernel_1355220/3035754114.py:7: PerformanceWarning: The following operation may generate 2610206368 cells in the resulting pandas object.\n",
      "  wide_df = long_df.pivot(index='title', columns='sourceID', values='reads').fillna(0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 2797 junctions that passed the filter.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sourceID</th>\n",
       "      <th>CGND-HRA-00013</th>\n",
       "      <th>CGND-HRA-00015</th>\n",
       "      <th>CGND-HRA-00017</th>\n",
       "      <th>CGND-HRA-00019</th>\n",
       "      <th>CGND-HRA-00020-2</th>\n",
       "      <th>CGND-HRA-00021</th>\n",
       "      <th>CGND-HRA-00023</th>\n",
       "      <th>CGND-HRA-00024</th>\n",
       "      <th>CGND-HRA-00025</th>\n",
       "      <th>CGND-HRA-00026</th>\n",
       "      <th>...</th>\n",
       "      <th>CGND-HRA-03135</th>\n",
       "      <th>CGND-HRA-03136</th>\n",
       "      <th>CGND-HRA-03137</th>\n",
       "      <th>CGND-HRA-03138</th>\n",
       "      <th>CGND-HRA-03139</th>\n",
       "      <th>CGND-HRA-03140</th>\n",
       "      <th>CGND-HRA-1927-1</th>\n",
       "      <th>CGND-HRA-1930-1</th>\n",
       "      <th>CGND-HRA-1941-1</th>\n",
       "      <th>CGND-HRA-1957-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>title</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10:119042185:119042215:-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:119042185:119042245:-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:119042185:119042275:-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:119042190:119042373:-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:119042215:119042638:-</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1876 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sourceID                     CGND-HRA-00013  CGND-HRA-00015  CGND-HRA-00017  \\\n",
       "title                                                                         \n",
       "chr10:119042185:119042215:-             0.0             0.0             0.0   \n",
       "chr10:119042185:119042245:-             0.0             0.0             0.0   \n",
       "chr10:119042185:119042275:-             0.0             0.0             0.0   \n",
       "chr10:119042190:119042373:-             0.0             0.0             0.0   \n",
       "chr10:119042215:119042638:-             0.0             0.0             0.0   \n",
       "\n",
       "sourceID                     CGND-HRA-00019  CGND-HRA-00020-2  CGND-HRA-00021  \\\n",
       "title                                                                           \n",
       "chr10:119042185:119042215:-             0.0               0.0             0.0   \n",
       "chr10:119042185:119042245:-             0.0               0.0             0.0   \n",
       "chr10:119042185:119042275:-             0.0               0.0             0.0   \n",
       "chr10:119042190:119042373:-             0.0               0.0             0.0   \n",
       "chr10:119042215:119042638:-             0.0               0.0             0.0   \n",
       "\n",
       "sourceID                     CGND-HRA-00023  CGND-HRA-00024  CGND-HRA-00025  \\\n",
       "title                                                                         \n",
       "chr10:119042185:119042215:-             0.0             0.0             0.0   \n",
       "chr10:119042185:119042245:-             0.0             0.0             0.0   \n",
       "chr10:119042185:119042275:-             0.0             0.0             0.0   \n",
       "chr10:119042190:119042373:-             0.0             0.0             0.0   \n",
       "chr10:119042215:119042638:-             0.0             0.0             0.0   \n",
       "\n",
       "sourceID                     CGND-HRA-00026  ...  CGND-HRA-03135  \\\n",
       "title                                        ...                   \n",
       "chr10:119042185:119042215:-             0.0  ...             0.0   \n",
       "chr10:119042185:119042245:-             0.0  ...             0.0   \n",
       "chr10:119042185:119042275:-             0.0  ...             0.0   \n",
       "chr10:119042190:119042373:-             0.0  ...             0.0   \n",
       "chr10:119042215:119042638:-             0.0  ...             0.0   \n",
       "\n",
       "sourceID                     CGND-HRA-03136  CGND-HRA-03137  CGND-HRA-03138  \\\n",
       "title                                                                         \n",
       "chr10:119042185:119042215:-             0.0             0.0             0.0   \n",
       "chr10:119042185:119042245:-             0.0             0.0             0.0   \n",
       "chr10:119042185:119042275:-             0.0             0.0             0.0   \n",
       "chr10:119042190:119042373:-             0.0             0.0             0.0   \n",
       "chr10:119042215:119042638:-             0.0             0.0             0.0   \n",
       "\n",
       "sourceID                     CGND-HRA-03139  CGND-HRA-03140  CGND-HRA-1927-1  \\\n",
       "title                                                                          \n",
       "chr10:119042185:119042215:-             0.0             0.0              0.0   \n",
       "chr10:119042185:119042245:-             0.0           142.0              0.0   \n",
       "chr10:119042185:119042275:-             0.0             0.0              0.0   \n",
       "chr10:119042190:119042373:-             0.0             0.0              0.0   \n",
       "chr10:119042215:119042638:-             0.0             0.0              0.0   \n",
       "\n",
       "sourceID                     CGND-HRA-1930-1  CGND-HRA-1941-1  CGND-HRA-1957-1  \n",
       "title                                                                           \n",
       "chr10:119042185:119042215:-              0.0              0.0              0.0  \n",
       "chr10:119042185:119042245:-              0.0              0.0              0.0  \n",
       "chr10:119042185:119042275:-              0.0              0.0              0.0  \n",
       "chr10:119042190:119042373:-              0.0              0.0              0.0  \n",
       "chr10:119042215:119042638:-             30.0              0.0              0.0  \n",
       "\n",
       "[5 rows x 1876 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter exitron data\n",
    "filtered_exitron_data = filterExitronData('exitron_data.parquet')\n",
    "filtered_exitron_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6. Normalization "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalizeExitronData(filtered_exitron_data, output_filepath):\n",
    "    annots = pd.DataFrame({\"exitron\": list(filtered_exitron_data.index)})\n",
    "    annots = pd.concat([annots, annots['exitron'].str.split(':', expand=True)], axis=1)\n",
    "    annots.columns = [\"exitron\", \"chr\", \"start\", \"end\", \"strand\"]\n",
    "    annots['start'] = annots['start'].astype(int)\n",
    "    annots['end'] = annots['end'].astype(int)\n",
    "\n",
    "    EXTEND = 6\n",
    "    normalized_df = np.zeros(filtered_exitron_data.shape)\n",
    "    ##tqdm(range(len(filtered_exitron_data.columns))):\n",
    "    for i in range(len(filtered_exitron_data.columns)):\n",
    "        print(\"Doing\",i)\n",
    "        source_id = filtered_exitron_data.columns[i]\n",
    "        bam_filepath = f'/gpfs/commons/projects/ALS_Consortium_resource/bam_files/{source_id}.bam'\n",
    "\n",
    "        with pysam.AlignmentFile(bam_filepath, 'rb') as bam:\n",
    "            for j in range(len(filtered_exitron_data.index)):\n",
    "                numerator = filtered_exitron_data.iat[j, i]\n",
    "\n",
    "                if numerator > 0:\n",
    "                    junc_chr = annots['chr'][j]\n",
    "                    junc_start = annots['start'][j]\n",
    "                    junc_end = annots['end'][j]\n",
    "\n",
    "                    # calculate denominator\n",
    "                    depth = bam.count_coverage(junc_chr, junc_start - EXTEND, junc_end + EXTEND, quality_threshold=0)\n",
    "                    totals = [sum(values) for values in zip(*depth)]\n",
    "                    \n",
    "                    denom = np.median(totals[:EXTEND] + totals[-EXTEND:])\n",
    "                    if denom > 0:\n",
    "                        normalized_df[j, i] = numerator / denom\n",
    "    np.save(output_filepath, normalized_df)\n",
    "    return normalized_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1876 [00:00<?, ?it/s]"
     ]
    }
   ],
   "source": [
    "normalizeExitronData(filtered_exitron_data, 'normalized_data.npy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7. Idenitfying Novel vs Annotated Exitrons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff = pr.read_gff3(\"gencode.v48.annotation.gff3.gz\")\n",
    "exons = gff[gff.Feature == \"exon\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findAnnotatedJunctions(exitrons, exons):\n",
    "\n",
    "    # parse IDs from  exon data\n",
    "    exons_df = exons.df\n",
    "    exons_df['transcript_id'] = exons_df['ID'].str.extract(r'exon:([^:]+):\\d+')\n",
    "\n",
    "    start_coord_to_transcripts = defaultdict(set)\n",
    "    end_coord_to_transcripts = defaultdict(set)\n",
    "\n",
    "    for row in exons_df.itertuples(index=False):\n",
    "        start_key = f\"{row.Chromosome}:{row.Start}\"\n",
    "        end_key = f\"{row.Chromosome}:{row.End}\"\n",
    "        \n",
    "        start_coord_to_transcripts[start_key].add(row.transcript_id)\n",
    "        end_coord_to_transcripts[end_key].add(row.transcript_id)\n",
    "\n",
    "    # parse exitron junctions\n",
    "    split_coords = exitrons.index.str.split(':').tolist()\n",
    "    junctions_long = pd.DataFrame(\n",
    "        split_coords,\n",
    "        columns=['Chromosome', 'Start', 'End', 'Strand'],\n",
    "        index=exitrons.index\n",
    "    )\n",
    "    junctions_long['Start'] = pd.to_numeric(junctions_long['Start'])\n",
    "    junctions_long['End'] = pd.to_numeric(junctions_long['End'])\n",
    "\n",
    "    # find matches for each junction\n",
    "    results_list = []\n",
    "    for junc in junctions_long.itertuples():\n",
    "        # exitron start should match an exon's end, exitron end should match exon's start\n",
    "        junc_start_key = f\"{junc.Chromosome}:{junc.Start}\"\n",
    "        junc_end_key = f\"{junc.Chromosome}:{junc.End}\"\n",
    "        \n",
    "        transcripts_at_junc_start = end_coord_to_transcripts.get(junc_start_key, set())\n",
    "        transcripts_at_junc_end = start_coord_to_transcripts.get(junc_end_key, set())\n",
    "        \n",
    "        # find the intersection of the two sets\n",
    "        common_transcripts = transcripts_at_junc_start.intersection(transcripts_at_junc_end)\n",
    "\n",
    "        results_list.append({\n",
    "            'start_match': bool(transcripts_at_junc_start),\n",
    "            'end_match': bool(transcripts_at_junc_end),\n",
    "            'transcript_match': bool(common_transcripts),\n",
    "            'start_transcript': \", \".join(sorted(list(transcripts_at_junc_start))),\n",
    "            'end_transcript': \", \".join(sorted(list(transcripts_at_junc_end))),\n",
    "            'common_transcripts': \", \".join(sorted(list(common_transcripts)))\n",
    "        })\n",
    "\n",
    "    return pd.DataFrame(results_list, index=exitrons.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_match</th>\n",
       "      <th>end_match</th>\n",
       "      <th>transcript_match</th>\n",
       "      <th>start_transcript</th>\n",
       "      <th>end_transcript</th>\n",
       "      <th>common_transcripts</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>title</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10:119042185:119042215:-</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:119042185:119042245:-</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:119042185:119042275:-</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:119042190:119042373:-</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10:119042215:119042638:-</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX:76428919:76429027:+</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX:76428955:76429063:+</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX:77683451:77683538:-</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000624032.3, ENST00000624166.3</td>\n",
       "      <td>ENST00000624032.3, ENST00000624166.3</td>\n",
       "      <td>ENST00000624032.3, ENST00000624166.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX:93672617:93672659:-</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX:93672659:93672701:-</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2797 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             start_match  end_match  transcript_match  \\\n",
       "title                                                                   \n",
       "chr10:119042185:119042215:-        False      False             False   \n",
       "chr10:119042185:119042245:-        False      False             False   \n",
       "chr10:119042185:119042275:-        False      False             False   \n",
       "chr10:119042190:119042373:-        False      False             False   \n",
       "chr10:119042215:119042638:-        False      False             False   \n",
       "...                                  ...        ...               ...   \n",
       "chrX:76428919:76429027:+           False      False             False   \n",
       "chrX:76428955:76429063:+           False      False             False   \n",
       "chrX:77683451:77683538:-            True       True              True   \n",
       "chrX:93672617:93672659:-           False      False             False   \n",
       "chrX:93672659:93672701:-           False      False             False   \n",
       "\n",
       "                                                 start_transcript  \\\n",
       "title                                                               \n",
       "chr10:119042185:119042215:-                                         \n",
       "chr10:119042185:119042245:-                                         \n",
       "chr10:119042185:119042275:-                                         \n",
       "chr10:119042190:119042373:-                                         \n",
       "chr10:119042215:119042638:-                                         \n",
       "...                                                           ...   \n",
       "chrX:76428919:76429027:+                                            \n",
       "chrX:76428955:76429063:+                                            \n",
       "chrX:77683451:77683538:-     ENST00000624032.3, ENST00000624166.3   \n",
       "chrX:93672617:93672659:-                                            \n",
       "chrX:93672659:93672701:-                                            \n",
       "\n",
       "                                                   end_transcript  \\\n",
       "title                                                               \n",
       "chr10:119042185:119042215:-                                         \n",
       "chr10:119042185:119042245:-                                         \n",
       "chr10:119042185:119042275:-                                         \n",
       "chr10:119042190:119042373:-                                         \n",
       "chr10:119042215:119042638:-                                         \n",
       "...                                                           ...   \n",
       "chrX:76428919:76429027:+                                            \n",
       "chrX:76428955:76429063:+                                            \n",
       "chrX:77683451:77683538:-     ENST00000624032.3, ENST00000624166.3   \n",
       "chrX:93672617:93672659:-                                            \n",
       "chrX:93672659:93672701:-                                            \n",
       "\n",
       "                                               common_transcripts  \n",
       "title                                                              \n",
       "chr10:119042185:119042215:-                                        \n",
       "chr10:119042185:119042245:-                                        \n",
       "chr10:119042185:119042275:-                                        \n",
       "chr10:119042190:119042373:-                                        \n",
       "chr10:119042215:119042638:-                                        \n",
       "...                                                           ...  \n",
       "chrX:76428919:76429027:+                                           \n",
       "chrX:76428955:76429063:+                                           \n",
       "chrX:77683451:77683538:-     ENST00000624032.3, ENST00000624166.3  \n",
       "chrX:93672617:93672659:-                                           \n",
       "chrX:93672659:93672701:-                                           \n",
       "\n",
       "[2797 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "findAnnotatedJunctions(filtered_exitron_data, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_results = findAnnotatedJunctions(filtered_exitron_data, exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proportion of annotated junctions by strand:\n",
      "Strand\n",
      "+    0.168254\n",
      "-    0.082613\n",
      "Name: transcript_match, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# finding proportion of annotated junctions by strand\n",
    "final_results['Strand'] = filtered_exitron_data.index.str[-1]\n",
    "\n",
    "if '+' in final_results['Strand'].values and '-' in final_results['Strand'].values:\n",
    "    strand_proportions = final_results.groupby('Strand')['transcript_match'].mean()\n",
    "\n",
    "    print(\"Proportion of annotated junctions by strand:\")\n",
    "    print(strand_proportions)\n",
    "\n",
    "else:\n",
    "    print(\"Could not perform strand comparison: Data for both '+' and '-' strands was not found in the results.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Junctions where boundaries match but transcripts do not:\n",
      "                             start_match  end_match  transcript_match  \\\n",
      "title                                                                   \n",
      "chr11:33542146:33542920:+           True      False             False   \n",
      "chr11:6533806:6533896:+             True      False             False   \n",
      "chr12:124911899:124913495:-        False       True             False   \n",
      "chr12:124911952:124913320:-        False       True             False   \n",
      "chr12:124912180:124912583:-         True      False             False   \n",
      "chr12:124912180:124912636:-         True      False             False   \n",
      "chr12:124912180:124913092:-         True      False             False   \n",
      "chr12:124912180:124913548:-         True      False             False   \n",
      "chr12:124912408:124913320:-        False       True             False   \n",
      "chr12:124912583:124913495:-        False       True             False   \n",
      "chr12:124912636:124913320:-        False       True             False   \n",
      "chr12:124912962:124913320:-        False       True             False   \n",
      "chr12:124913039:124913495:-        False       True             False   \n",
      "chr12:124913092:124913320:-        False       True             False   \n",
      "chr12:124913190:124913320:-        False       True             False   \n",
      "chr12:124913267:124913723:-         True      False             False   \n",
      "chr12:53985347:53985484:+          False       True             False   \n",
      "chr15:73740472:73740567:-          False       True             False   \n",
      "chr16:728868:729008:+              False       True             False   \n",
      "chr17:16382111:16382567:+          False       True             False   \n",
      "chr17:16382130:16382411:+           True      False             False   \n",
      "chr17:16382339:16382567:+          False       True             False   \n",
      "chr17:81120571:81120827:-          False       True             False   \n",
      "chr17:81120937:81121247:-          False       True             False   \n",
      "chr19:3054037:3054119:-            False       True             False   \n",
      "chr19:44908944:44908974:+           True      False             False   \n",
      "chr19:4792013:4792139:+             True      False             False   \n",
      "chr20:3165670:3165766:-            False       True             False   \n",
      "chr20:63521589:63521704:+           True      False             False   \n",
      "chr20:63941043:63941135:-           True      False             False   \n",
      "chr21:45456176:45456266:+           True      False             False   \n",
      "chr6:3154487:3154519:-              True      False             False   \n",
      "chr6:89080973:89081078:+            True      False             False   \n",
      "chr8:13006356:13006385:+            True      False             False   \n",
      "chr8:24914717:24914810:+            True      False             False   \n",
      "chr8:27605017:27605171:-           False       True             False   \n",
      "chr8:27605149:27605171:-           False       True             False   \n",
      "chr9:128178988:128179072:-         False       True             False   \n",
      "chr9:128250767:128250899:+         False       True             False   \n",
      "chr9:128250848:128250899:+         False       True             False   \n",
      "\n",
      "                                                              start_transcript  \\\n",
      "title                                                                            \n",
      "chr11:33542146:33542920:+                                    ENST00000526400.7   \n",
      "chr11:6533806:6533896:+                                      ENST00000534210.6   \n",
      "chr12:124911899:124913495:-                                                      \n",
      "chr12:124911952:124913320:-                                                      \n",
      "chr12:124912180:124912583:-                                  ENST00000538617.5   \n",
      "chr12:124912180:124912636:-                                  ENST00000538617.5   \n",
      "chr12:124912180:124913092:-                                  ENST00000538617.5   \n",
      "chr12:124912180:124913548:-                                  ENST00000538617.5   \n",
      "chr12:124912408:124913320:-                                                      \n",
      "chr12:124912583:124913495:-                                                      \n",
      "chr12:124912636:124913320:-                                                      \n",
      "chr12:124912962:124913320:-                                                      \n",
      "chr12:124913039:124913495:-                                                      \n",
      "chr12:124913092:124913320:-                                                      \n",
      "chr12:124913190:124913320:-                                                      \n",
      "chr12:124913267:124913723:-                                  ENST00000541272.1   \n",
      "chr12:53985347:53985484:+                                                        \n",
      "chr15:73740472:73740567:-                                                        \n",
      "chr16:728868:729008:+                                                            \n",
      "chr17:16382111:16382567:+                                                        \n",
      "chr17:16382130:16382411:+                                    ENST00000535788.1   \n",
      "chr17:16382339:16382567:+                                                        \n",
      "chr17:81120571:81120827:-                                                        \n",
      "chr17:81120937:81121247:-                                                        \n",
      "chr19:3054037:3054119:-                                                          \n",
      "chr19:44908944:44908974:+                                    ENST00000446996.5   \n",
      "chr19:4792013:4792139:+                                      ENST00000601192.1   \n",
      "chr20:3165670:3165766:-                                                          \n",
      "chr20:63521589:63521704:+    ENST00000370179.8, ENST00000473620.1, ENST0000...   \n",
      "chr20:63941043:63941135:-    ENST00000354216.11, ENST00000369908.9, ENST000...   \n",
      "chr21:45456176:45456266:+                                   ENST00000355480.10   \n",
      "chr6:3154487:3154519:-                                       ENST00000842424.1   \n",
      "chr6:89080973:89081078:+                                     ENST00000369472.1   \n",
      "chr8:13006356:13006385:+     ENST00000447063.6, ENST00000524591.7, ENST0000...   \n",
      "chr8:24914717:24914810:+                                     ENST00000607058.3   \n",
      "chr8:27605017:27605171:-                                                         \n",
      "chr8:27605149:27605171:-                                                         \n",
      "chr9:128178988:128179072:-                                                       \n",
      "chr9:128250767:128250899:+                                                       \n",
      "chr9:128250848:128250899:+                                                       \n",
      "\n",
      "                                                                end_transcript  \n",
      "title                                                                           \n",
      "chr11:33542146:33542920:+                                                       \n",
      "chr11:6533806:6533896:+                                                         \n",
      "chr12:124911899:124913495:-                                  ENST00000541272.1  \n",
      "chr12:124911952:124913320:-                                  ENST00000538617.5  \n",
      "chr12:124912180:124912583:-                                                     \n",
      "chr12:124912180:124912636:-                                                     \n",
      "chr12:124912180:124913092:-                                                     \n",
      "chr12:124912180:124913548:-                                                     \n",
      "chr12:124912408:124913320:-                                  ENST00000538617.5  \n",
      "chr12:124912583:124913495:-                                  ENST00000541272.1  \n",
      "chr12:124912636:124913320:-                                  ENST00000538617.5  \n",
      "chr12:124912962:124913320:-                                  ENST00000538617.5  \n",
      "chr12:124913039:124913495:-                                  ENST00000541272.1  \n",
      "chr12:124913092:124913320:-                                  ENST00000538617.5  \n",
      "chr12:124913190:124913320:-                                  ENST00000538617.5  \n",
      "chr12:124913267:124913723:-                                                     \n",
      "chr12:53985347:53985484:+                                    ENST00000515593.1  \n",
      "chr15:73740472:73740567:-                                    ENST00000558834.1  \n",
      "chr16:728868:729008:+        ENST00000389701.9, ENST00000389703.8, ENST0000...  \n",
      "chr17:16382111:16382567:+                                    ENST00000583934.1  \n",
      "chr17:16382130:16382411:+                                                       \n",
      "chr17:16382339:16382567:+                                    ENST00000583934.1  \n",
      "chr17:81120571:81120827:-                                    ENST00000570932.5  \n",
      "chr17:81120937:81121247:-                 ENST00000374792.6, ENST00000570932.5  \n",
      "chr19:3054037:3054119:-      ENST00000221561.12, ENST00000327141.9, ENST000...  \n",
      "chr19:44908944:44908974:+                                                       \n",
      "chr19:4792013:4792139:+                                                         \n",
      "chr20:3165670:3165766:-                   ENST00000360342.7, ENST00000645462.1  \n",
      "chr20:63521589:63521704:+                                                       \n",
      "chr20:63941043:63941135:-                                                       \n",
      "chr21:45456176:45456266:+                                                       \n",
      "chr6:3154487:3154519:-                                                          \n",
      "chr6:89080973:89081078:+                                                        \n",
      "chr8:13006356:13006385:+                                                        \n",
      "chr8:24914717:24914810:+                                                        \n",
      "chr8:27605017:27605171:-                                     ENST00000523589.5  \n",
      "chr8:27605149:27605171:-                                     ENST00000523589.5  \n",
      "chr9:128178988:128179072:-   ENST00000372948.7, ENST00000372954.5, ENST0000...  \n",
      "chr9:128250767:128250899:+                                   ENST00000706054.1  \n",
      "chr9:128250848:128250899:+                                   ENST00000706054.1  \n",
      "Number of exitrons without annotations: 40\n"
     ]
    }
   ],
   "source": [
    "mismatched_transcripts = final_results[\n",
    "    (final_results['transcript_match'] == False) & \n",
    "    (final_results['start_match'] | final_results['end_match'])\n",
    "]\n",
    "\n",
    "print(\"Junctions where boundaries match but transcripts do not:\")\n",
    "print(mismatched_transcripts)\n",
    "print(f\"Number of exitrons without annotations: {len(mismatched_transcripts)}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vscode",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
